{% extends "portal/compose_base.html" %}
{% comment %}
  Upload Sequences
{% endcomment %}

{% block javascript %}

<script type="text/javascript">
	var countUserMsas = 0;

	function addMsaLi() {
			
		countUserMsas += 1;
		
		var new_idbase = "user_msa";		
		var new_id = new_idbase.concat(  countUserMsas.toString()  );
		
		var parentp = document.getElementById("usermsas_p");

		var new_childp = document.createElement("p");
		var new_childpid = "childp".concat( new_id );
		new_childp.setAttribute("id", new_childpid);
		
		var label = document.createElement("label");
		label.setAttribute("for", new_id);
		
		var textbox = document.createElement("input");
		textbox.setAttribute("id", 	.concat("_name") ); 
		textbox.setAttribute("name", new_id.concat("_name") );
		textbox.setAttribute("type", "Text");
		textbox.setAttribute("value", "my alignment".concat(countUserMsas) );
		textbox.setAttribute("size", "13");
		
		var fileinput = document.createElement("input");
		fileinput.setAttribute("id", new_id.concat("_file")  );
		fileinput.setAttribute("name", new_id.concat("_file"));
		fileinput.setAttribute("type", "file");
		fileinput.setAttribute("size", "8");
		
		var removelink = document.createElement("a");
		removelink.setAttribute("id", "removeusermsalink");
		removelink.setAttribute("title", "Remove this user-defined sequence alignment");
		removelink.setAttribute("href", "PleaseEnableJavascript.html");
		var onclickstring = "removeMsaLi(\"";
		onclickstring = onclickstring.concat(new_childpid);
		onclickstring = onclickstring.concat("\"); return false;");
		removelink.setAttribute("onclick", onclickstring );
		removelink.innerHTML = "(remove)";
		removelink.setAttribute("class", "smalltext");
		
		new_childp.appendChild(label);
		new_childp.appendChild(textbox);
		new_childp.appendChild(fileinput);
		new_childp.appendChild(removelink);
		parentp.appendChild(new_childp);
	}
	
	function removeMsaLi(new_childpid) {
		//alert("opened removeMsaLi");
		var this_ele = document.getElementById(new_childpid);
		this_ele.parentNode.removeChild(this_ele);		
	}
</script>

{% endblock %}

{% block compose_content %}
{% load staticfiles %}

<form id="user_form" method="post" action="/portal/compose1" enctype="multipart/form-data">
{% csrf_token %}

<h2>Step 1: Setup the Basics</h2>

{% for msg in error_messages %}
{% if msg != None %}
<div class="rounded_div_medium" style="background-color:pink;">
<h3 class="red">A problem occurred. . .</h3>
<p>&#10140; {{msg}}</p>
</div>
{% endif %}
{% endfor %}

<div style="float:left; width:49%">
	<div class="rounded_div_medium" style="background-color:#dedede;">
	    <table width="100%">
	    <tr>
	    <td>
	    <h3>Protein Sequences
	     	<a href="#" class="tooltip">
			<img src="{% static "phylobot/question_icon.png" %}" height="15px">
			<span>
	        	<p class="smalltext">Select a FASTA-formatted file containing orthologous amino acid sequences
	        	for your protein family of interest. These sequences do not need to be aligned.</p>
	    	</span>
			</a>
	    </h3>
	    </td>

	    <td class="smalltext" align="right">
	    <a href="/faq_fasta">FASTA Help</a>
	    </td>
	    </tr>
	    </table>
		{% if aa_seqfile_url %}
	    	<p>Current File: <a href="{% static aa_seqfile_url %}">{{aa_seqfile_short}}</a></p>
	    {% endif %}
	    {{ aa_seqfileform.as_p }}
		<p class="smalltext"><input id="aa_is_uniprot" name="is_uniprot" type="checkbox" value="aa"/>Click here if your sequence names are formatted using the <a href="http://www.uniprot.org/help/fasta-headers">NCBI/UniProtKB format.</a></p>
	   
	
	</div>

	<!--
	<div class="rounded_div_medium" style="background-color:#dedede;">
	    <table width="100%">
	    <tr>
	    <td>
	    <h3>Codon Sequences (optional)
	     <a href="#" class="tooltip">
			<img src="{% static "phylobot/question_icon.png" %}" height="15px">
			<span>
	        	<p class="smalltext">PhyloBot can use codon sequences to calculate dN/dS tests of positive selection.
	        	Select a FASTA-formatted file containing orthologous codon sequences
	        	for your protein family of interest. Ensure the sequences use the same names
	        	as your protein sequences, and ensure you provide codon content for every amino
	        	acid in the proteins.</p>
	    	</span>
		</a>
	    </h3>
	    </td>

	    <td class="smalltext" align="right">
	    <a href="/faq_fasta">FASTA Help</a>
	    </td>
	    </tr>
	    </table>

	    {% if codon_seqfile_url %}
	    	<p>Current File: <a href="{% static codon_seqfile_url %}">{{codon_seqfile_short}}</a></p>
	    {% endif %}
	    {{ codon_seqfileform.as_p }}
		<p class="smalltext"><input id="nt_is_uniprot" name="is_uniprot" type="checkbox" value="nt"/>Click here if your sequence names are formatted using the <a href="http://www.uniprot.org/help/fasta-headers">NCBI/UniProtKB format.</a></p>
	   	
	</div>
	-->
	
	<div class="rounded_div_medium" style="background-color:#dedede;">
		    <h3>Job Name</h3>
		<p>Choose a Short Name:{{ js_form.name }}
		<p>
		Describe this project (optional):
		{{ js_form.project_description }}</p>
		<p>
	</div>
</div>

<div style="margin-left:50%; width:50%">
	<div class="rounded_div_medium" style="background-color:#dedede;">
		<h3>Sequence Alignment Algorithms
	     <a href="#" class="tooltip">
			<img src="{% static "phylobot/question_icon.png" %}" height="15px">
			<span>
	        	<p class="smalltext">The selected algorithms will be used to align your orthologous sequences. 
	        	Select one, many, or all of these algorithms.</p>
	    	</span>
		</a>
		</h3>
		
		{{ js_form.alignment_algorithms }}
		
		<!--
		The User-spec alignments
		-->
				
		<h3>Custom Sequence Alignments (optional)
	     <a href="#" class="tooltip">
			<img src="{% static "phylobot/question_icon.png" %}" height="15px">
			<span>
	        	<p class="smalltext">Upload FASTA-formatted sequence alignments. These alignments will be
	        	analyzed alongside alignments computed by the algorithms in PhyloBot.</p>
	    	</span>
		</a>
		</h3>
        <p id="usermsas_p">
        	<!--
        	<table>
	            {% for usermsa in user_msas %}			
					<tr></td>
					<p id="childp{{forloop.counter}}">
					<label for="childp{{forloop.counter}}"></label>
					<input type="text" id="user_msa{{forloop.counter}}_name" name="user_msa{{forloop.counter}}_name" value="{{usermsa.0}}" size="13"/>
					<input type="file" id="user_msa{{forloop.counter}}_file" name="user{{forloop.counter}}_file" size="1"/>
					</td>
					<td width="50%">
					<a class="smalltext" href="PleaseEnableJavascript.html;" onclickstring="removeMsaLi('childp{{forloop.counter}}'); return false;">(remove)</a>
					<a href="{{usermsa.1}}" class="smalltext">(download)</a>
					</p>
		        	</td></tr>
	        	{% endfor %}
        	</table>			
        	-->
        </p>
        
		<p><a id="addusermsalink" title="+ Add a user-defined sequence alignment" href="PleaseEnableJavascript.html" onclick="addMsaLi(); return false;">Add a user-defined sequence alignment</a>
		</p>
		
		<p>
		<h3>Phylogenetic Models
	     <a href="#" class="tooltip">
			<img src="{% static "phylobot/question_icon.png" %}" height="15px">
			<span>
	        	<p class="smalltext">The selected phylogenetic models will be used to infer maximum likelihood
	        		phylogenies for your orthologous sequences. Select one, many, or all of these models.</p>
	    	</span>
		</a>
		</h3>
		{{ js_form.raxml_models }}
		<p>
	</div>
	
	<div class="rounded_div_medium" style="background-color:#dedede;">
	   	<table width="100%">
	    <tr>
	    <td>
	    <h3>Constraint Tree (optional)
	     	<a href="#" class="tooltip">
			<img src="{% static "phylobot/question_icon.png" %}" height="15px">
			<span>
	        	<p class="smalltext">You can constrain the phylogenetic inference to always include certain
	        	branching relationships that you know <em>a priori</em> to be true. The constraint tree
	        	is a Newick-formatted file that describes the relationship between three or more of your
	        	protein sequences. The constraint tree does not need to include a reference for all
	        	sequences.</p>
	    	</span>
			</a>
	    </h3>
	    </td>
	    <td class="smalltext" align="right">
	    <a href="/faq_newick">Trees Help</a>
	    </td>
	    </tr>
	    </table>

	    {% if constrainttree_seqfile_url %}
	    	<p>Current File: <a href="{% static constrainttree_seqfile_url %}">{{constrainttree_seqfile_short}}</a></p>
	    {% endif %}
	    {{ constrainttree_seqfileform.as_p }}	
	</div>

	<div align="center">
	<button class="button1"  type="submit" name="submit" align="center">Save and Continue</button>
	</div>

    <!--<input type="submit" name="submit" value="Save and Continue" />-->
	    
</div>


</form>
{% endblock %}