{% extends "libview/libview_base.html" %}
{% load staticfiles %}  
{% block body %}

<table width=100%>
<tr>
	<th align='left'>ID</th>
	<th align='left'>Alignment
			<a href="#" class="tooltip">
			<img src="{% static "phylobot/question_icon.png" %}" height="15px">
			<span align="left">
				<h3>Sequence Alignment Method</h3>
	        	<p class="smalltext">Alignments computed by different methods can disagree about which sites are homologous. These differences can, potentially, impact the topology and branch lengths on the maximum likelihood phylogeny.</p>
	    	</span>
	    	</a>
	</th>

	<th align='center'>Model
			<a href="#" class="tooltip">
			<img src="{% static "phylobot/question_icon.png" %}" height="15px">
			<span align="left">
				<h3>Markov Model</h3>
	        	<p class="smalltext">Each model specifies a different substitution pattern between amino acids. These differences can, potentially, lead to different inferences of the best tree topology and branch lengths.</p>
	    	</span>
	    	</a>
	</th>
	<th align='center'>log(L)
			<a href="#" class="tooltip">
			<img src="{% static "phylobot/question_icon.png" %}" height="15px">
			<span align="left">
				<h3>log(Likelihood)</h3>
	        	<p class="smalltext">The log of the likelihood, i.e. the probability of observing all the data in the sequence alignment, given the evolutionary model, the best tree topology, and optimized branch lengths on that tree.</p>
	    	</span>
	    	</a>
	</th>
	<th align='center'>Relative<br>Prob.
			<a href="#" class="tooltip">
			<img src="{% static "phylobot/question_icon.png" %}" height="15px">
			<span align="left">
				<h3>Relative Probability</h3>
	        	<p class="smalltext">The relatively probability of the phylogeny inferred by this alignment and model, compared to the best phylogenies from other models with this same alignment. The relative probability is computed based on the log(likelihood).</p>
	    	</span>
	    	</a>
	</th>
	<th align='center'>Alpha
			<a href="#" class="tooltip">
			<img src="{% static "phylobot/question_icon.png" %}" height="15px">
			<span align="left">
				<h3>Alpha</h3>
	        	<p class="smalltext">The alpha parameter describes the scale of the distribution of branch lengths. A high value for alpha generally indicates long branches.</p>
	    	</span>
	    	</a>
	</th>
	<th align='center'>&sum; B.L.
			<a href="#" class="tooltip">
			<img src="{% static "phylobot/question_icon.png" %}" height="15px">
			<span align="left">
				<h3>Sum of Branch Lengths</h3>
	        	<p class="smalltext">The sum of all branch lengths on the phylogeny.</p>
	    	</span>
	    	</a>
	</th>
	<th align='center'>Download
			<a href="#" class="tooltip">
			<img src="{% static "phylobot/question_icon.png" %}" height="15px">
			<span align="left">
				<h3>Download</h3>
	        	<p class="smalltext">You can download the tree as a Newick-formatted string, viewable by external programs such as FigTree.</p>
	    	</span>
	    	</a>
	</th>
	<th align='center'>Statistical<br>Support
			<a href="#" class="tooltip">
			<img src="{% static "phylobot/question_icon.png" %}" height="15px">
			<span align="left">
				<h3>Statistical Support</h3>
	        	<p class="smalltext">One or more methods were used to estimate the support for each branch in the phylogeny.</p>
	        	<p class="smalltext">1. The Approximate Likelihood Ratio Test (aLRT)</p>
	        	<p class="tinytext">
	        	Anisimova and Gascuel<br>
	        	<em>Approximate Likelihood-Ratio Test for Branches: A Fast, Accurate, and Powerful Alternative</em><br>
	        	Systematic Biology (2006) 55 (4): 539-552<br>
	        	DOI: 10.1080/10635150600755453
	        	</p>
	        	<p class="smalltext">2. The Approximate Likelihood Ratio (aLR), computed from aLRTs. INTERPRETATION: An aLR value of X on branch B indicates that the existence of B is X times more likely than the next-best phylogenetic hypothesis in which B does not exist.</p>
	    	</span>
	    	</a>
	</th>
	
	<th align='center'>Ancestors</th>
</tr>

{% for tuple in tree_tuples %}
<tr>
	<td>{{ forloop.counter0 }}.</td>
	<td>{{tuple.0}}</td>
	<td class='smalltext'>{{tuple.1}}</td>
	<td align='center'>{{tuple.2|floatformat:2}}</td>
	<td align='center'>{{tuple.3|floatformat:1}}</td>
	<td align='center'>{{tuple.4|floatformat:3}}</td>
	<td align='center'>{{tuple.5|floatformat:2}}</td>
	<td align='center'><a href='/{{alid}}/{{shortname}}.{{tuple.0}}.{{tuple.1}}.newick'>newick</a></td>
	<td align='center'>
	{% for m in support_methods %}
		<a href='/{{alid}}/{{shortname}}.{{tuple.0}}.{{tuple.1}}.{{m}}.newick'>{{m}}</a>
		{% if m != lastmethod %}
		|
		{% endif %}
	{% endfor %}
	</td>
	<td align='center'><a href="/{{alid}}/{{tuple.0}}.{{tuple.1}}/ancestors">view</a></td>
</tr>
{% endfor %}

</table>

<hr>
<h3>Structural Differences Between Trees		
		<a href="#" class="tooltip">
			<img src="{% static "phylobot/question_icon.png" %}" height="15px">
			<span>
	        	<p class="smalltext">The values in this table express symmmetric
	        	distance between trees inferred from different combinations of methods and models. 
	        	The IDs correspond to the row IDs in the table at top.
	        	Symmetric distance is computed as the sum of branch splits found in one
	        	tree, but not the other tree. For more information, see the DendroPy function 'symmetric_difference'.</p>
	    	</span>
		</a>

</h3>

<table width="50%">
<tr><td align="center">ID</td>
{% for item in symmd_matrix.0 %}
<td align="center">{{ forloop.counter0 }}</td>
{% endfor %}
</tr>
{% for row in symmd_matrix %}
	<tr align="center">
	<td align="center">{{ forloop.counter0 }}</td>
	{% for item in row %}
		<td		
		{% if item == symmd_matrix_colorbins.0 %}class="ppfull"
		{% elif item <= symmd_matrix_colorbins.1 %}class="pp9"
		{% elif item <= symmd_matrix_colorbins.2 %}class="pp8"
		{% elif item <= symmd_matrix_colorbins.3 %}class="pp7"
		{% elif item <= symmd_matrix_colorbins.4 %}class="pp6"
		{% elif item <= symmd_matrix_colorbins.5 %}class="pp5"
		{% else%}class="pplow"{% endif %}
		>{{ item|floatformat:2 }}</td>
	{% endfor %}
	</tr>
{% endfor %}
</table>

<hr>
<h3>Branch Length Differences Between Trees		
		<a href="#" class="tooltip">
			<img src="{% static "phylobot/question_icon.png" %}" height="15px">
			<span>
	        	<p class="smalltext">The values in this table express the Euclidean
	        	distance between trees inferred from different combinations of 
	        	methods and model. The IDs correspond to the row IDs in the table at top.
	        	Euclidean distance is the sum of absolute differences in branch lengths
	        	for equivalent splits between two trees, with a value of 0.0 for
	        	the branch length of a missing split. This is equivalent to the "branch length distance"
	        	proposed by Felsenstein (2004). For more information, see the DendroPy function 'euclidean_distance'.</p>
	    	</span>
		</a>

</h3>

<table width="50%">
<tr><td align="center">ID</td>
{% for item in eucd_matrix.0 %}
<td align="center">{{ forloop.counter0 }}</td>
{% endfor %}
</tr>
{% for row in eucd_matrix %}
	<tr align="center">
	<td align="center">{{ forloop.counter0 }}</td>
	{% for item in row %}
		<td		
		{% if item == eucd_matrix_colorbins.0 %}class="ppfull"
		{% elif item <= eucd_matrix_colorbins.1 %}class="pp9"
		{% elif item <= eucd_matrix_colorbins.2 %}class="pp8"
		{% elif item <= eucd_matrix_colorbins.3 %}class="pp7"
		{% elif item <= eucd_matrix_colorbins.4 %}class="pp6"
		{% elif item <= eucd_matrix_colorbins.5 %}class="pp5"
		{% else%}class="pplow"{% endif %}
		>{{ item|floatformat:2 }}</td>
	{% endfor %}
	</tr>
{% endfor %}
</table>


{% endblock %}