{% extends "libview/libview_base.html" %}

{% block head %}
<script type="text/javascript" src="https://www.google.com/jsapi"></script><script type="text/javascript">
google.load("visualization", "1.1", {packages:["corechart", "line"]});
google.setOnLoadCallback(drawChart);

function drawChart() {

	{% for data in zorro_tuples %}

		var data{{data.0}} = google.visualization.arrayToDataTable([
		['Site', 'ZORRO Score', {role:'style'}],
		{% for tuple in data.2 %}
			{% if tuple.0 != data.0 %}	[ {{tuple.0}}, {{tuple.1}}, '#1E90FF'],
			{% else %} [ {{tuple.0}}, {{tuple.1}}, '#1E90FF']
			{% endif %}
		{% endfor %}
		]);
		
		var options{{data.0}} = {
		  title: '{{data.0}}',
		  titleTextStyle: {fontSize:18, color: '#444444'},
		  textStyle:{fontSize: 11},
		  vAxis: {
		  	title: 'ZORRO Score',  
		  	titleTextStyle: {color: 'black', fontName:'Helvetica',fontSize:'12pt'}
		  	},
		  orientation: 'horizontal',
		  backgroundColor: '#f0f0f0',
		  bar:{groupWidth: '90%'},
		
		  hAxis: {
		      title: 'Sites in {{data.0}}',
		      textStyle:{fontSize:11}
		  },
		  vAxis: {
		      title: 'ZORRO Score',
		      textStyle:{fontSize:15}
		  },
		  legend: { position: "none" }
		};
		
		var chart{{data.0}} = new google.visualization.BarChart(document.getElementById('chart_div_{{data.0}}'));
		chart{{data.0}}.draw(data{{data.0}}, options{{data.0}});
	
	{% endfor %}

};

</script>
{% endblock %}


{% block body %}

{% load staticfiles %}

<table width=100%>
<tr><th align='left'>Alignment Method 
		<a href="#" class="tooltip">
		
			<img src="{% static "phylobot/question_icon.png" %}" height="15px">
			<span>
				<h3>Alignment Method</h3>
	        	<p class="smalltext">Each alignment method applies a unique approach to determine which sequence sites are homologous and the placement of gaps.</p><p class="smalltext">To learn more about a particular alignment method, click the link [cite] next to its name.</p>
	    	</span>
		</a>
	</th>	
	<th align='center'>N Taxa
		<a href="#" class="tooltip">
			<img src="{% static "phylobot/question_icon.png" %}" height="15px">
			<span align="left">
				<h3>N Taxa</h3>
	        	<p class="smalltext">The number of sequences aligned.</p>
	    	</span>
		</a>
	</th>
	<th align='center'>N Sites
			<a href="#" class="tooltip">
			<img src="{% static "phylobot/question_icon.png" %}" height="15px">
			<span align="left">
				<h3>N Sites</h3>
	        	<p class="smalltext">The number of sequence sites produced by this alignment method, including gaps. The number of sites in the alignment varies depending on the alignment method.</p>
	    	</span>
	    	</a>
	</th>
	<th align='center'>ZORRO Profile
			<a href="#" class="tooltip">
			<img src="{% static "phylobot/question_icon.png" %}" height="15px">
			<span align="left">
	        	<p class="smalltext">The ZORRO algorithm assigns a confidence score to each site in the alignment.
	        	Use the profile of ZORRO scores to determine the relative robustness of different regions of the alignment.</p>
	        	<p class="smalltext">Reference:</p>
	        	<p class="tinytext">
	        	Wu, Chatterji, and Eisen<br>
	        	<em>Accounting For Alignment Uncertainty in Phylogenomics</em><br>
	        	PLoS One, January 17, 2012<br>
	        	DOI: 10.1371/journal.pone.0030288
	        	</p>
	    	</span>
	    	</a>
	</th>
	<th align='center'>View All Sites</th>
	<th align='center'>Download</th>
</tr>

{% for tuple in msaid_tuples %}
<tr><td align='left'>{{tuple.0}} <font class="smalltext"><a href="{{tuple.3}}">[cite]</a></font></td>
	<td align='center'>{{tuple.1}}</td>
	<td align='center'>{{tuple.2}}</td>
	<td align='center'><a href='/{{alid}}/{{tuple.0}}/alignment'>launch viewer</a></td>
	<td align='center'>
		<a href='/{{alid}}/{{shortname}}.{{tuple.0}}.fasta'>fasta</a> |
		<a href='/{{alid}}/{{shortname}}.{{tuple.0}}.phylip'>phylip</a>
	</td>
	
</tr>
{% endfor %}
</table>

<hr>
<h3>Alignment Support Summary		
		<a href="#" class="tooltip">
			<img src="{% static "phylobot/question_icon.png" %}" height="15px">
			<span>
	        	<p class="smalltext">ZORRO scores were calculated for every site in every alignment.
	        	These scores estimate our confidence in assigning homology to a set of residues, versus
	        	another hypothesis in which those residues are aligned into two or more different sites.
	        	For more information, see Wu, Chatterji, and Eisen (PLoS One 2012).</p>
	    	</span>
		</a>

</h3>
<div align="center">
	{% for data in zorro_tuples %}
	<div id="chart_div_{{data.0}}" style="width: 100%; height: 200px;"></div>
	{% endfor %}
</div>

{% endblock %}